File: helpers/macos/Sources/SidecarKit/Protocol/SidecarFrame.swift
import Foundation

/// One message on the Sidecar protocol — a JSON object tagged with a `"type"` discriminator, framed
/// length-prefixed over the socket. The Swift mirror of `core/sidecar`'s sealed `SidecarFrame`.
///
/// Inbound (client → helper): `hello`, `request`, `cancel`.
/// Outbound (helper → client): `welcome`, `response`, `stream_data`, `stream_end`, `push`, `failure`.
///
/// Correlation `id`s are positive integers chosen by the client; the helper echoes them back.
public enum SidecarFrame: Equatable {
    case hello(token: String, protocolVersion: Int)
    case welcome(protocolVersion: Int, capabilities: [String])
    case request(id: Int64, method: String, params: JSONValue?)
    case response(id: Int64, result: JSONValue?)
    case streamData(id: Int64, event: JSONValue)
    case streamEnd(id: Int64)
    case cancel(id: Int64)
    case push(topic: String, payload: JSONValue)
    case failure(id: Int64?, error: SidecarError)
}

extension SidecarFrame {
    /// The `"type"` discriminator string (part of the contract).
    var type: String {
        switch self {
        case .hello: return "hello"
        case .welcome: return "welcome"
        case .request: return "request"
        case .response: return "response"
        case .streamData: return "stream_data"
        case .streamEnd: return "stream_end"
        case .cancel: return "cancel"
        case .push: return "push"
        case .failure: return "failure"
        }
    }

    /// The frame as an opaque object, **omitting null/default fields** to match the contract
    /// (`explicitNulls=false` / `encodeDefaults=false` on the JVM side): a null `params`/`result` and an
    /// empty `capabilities`/absent `id` are left off the wire.
    var jsonObject: JSONValue {
        var o: [String: JSONValue] = ["type": .string(type)]
        switch self {
        case .hello(let token, let protocolVersion):
            o["token"] = .string(token)
            o["protocolVersion"] = .int(Int64(protocolVersion))
        case .welcome(let protocolVersion, let capabilities):
            o["protocolVersion"] = .int(Int64(protocolVersion))
            if !capabilities.isEmpty { o["capabilities"] = .array(capabilities.map { .string($0) }) }
        case .request(let id, let method, let params):
            o["id"] = .int(id)
            o["method"] = .string(method)
            if let params { o["params"] = params }
        case .response(let id, let result):
            o["id"] = .int(id)
            if let result { o["result"] = result }
        case .streamData(let id, let event):
            o["id"] = .int(id)
            o["event"] = event
        case .streamEnd(let id):
            o["id"] = .int(id)
        case .cancel(let id):
            o["id"] = .int(id)
        case .push(let topic, let payload):
            o["topic"] = .string(topic)
            o["payload"] = payload
        case .failure(let id, let error):
            if let id { o["id"] = .int(id) }
            var err: [String: JSONValue] = [
                "code": .string(error.code.rawValue),
                "message": .string(error.message),
            ]
            if let details = error.details { err["details"] = details }
            o["error"] = .object(err)
        }
        return .object(o)
    }
}

extension SidecarFrame: CustomStringConvertible {
    /// **Privacy (ADR-0009):** payload-bearing frames redact their contents — only metadata (type, id,
    /// method, topic, error code) appears, so a stray log can never leak Transcript text.
    public var description: String {
        switch self {
        case .hello(_, let v): return "hello(token=<redacted>, protocolVersion=\(v))"
        case .welcome(let v, let caps): return "welcome(protocolVersion=\(v), capabilities=\(caps))"
        case .request(let id, let m, _): return "request(id=\(id), method=\(m), params=<redacted>)"
        case .response(let id, _): return "response(id=\(id), result=<redacted>)"
        case .streamData(let id, _): return "stream_data(id=\(id), event=<redacted>)"
        case .streamEnd(let id): return "stream_end(id=\(id))"
        case .cancel(let id): return "cancel(id=\(id))"
        case .push(let topic, _): return "push(topic=\(topic), payload=<redacted>)"
        case .failure(let id, let e): return "failure(id=\(id.map(String.init) ?? "nil"), code=\(e.code.rawValue))"
        }
    }
}

File: core/sidecar/src/main/kotlin/com/circuitstitch/deferno/core/sidecar/SidecarFrame.kt
package com.circuitstitch.deferno.core.sidecar

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * One message on the Sidecar [[Sidecar protocol]] — a JSON object framed length-prefixed over the
 * socket (ADR-0024). A closed (sealed) polymorphic hierarchy: every frame carries a `"type"`
 * discriminator (see [SidecarJson]). The three traffic shapes the contract must cover all reduce to
 * these frames:
 *
 * - **request/response** — client [Request] → server [Response] (or [Failure] with the same `id`);
 * - **server→client stream** — client [Request] (a stream-opening method) → server [StreamData]* then
 *   [StreamEnd] (or [Failure]); the client may [Cancel] early;
 * - **unsolicited push** — server [Push] with no correlation id.
 *
 * Plus the peer-auth handshake: client [Hello] → server [Welcome] (or [Failure] with `id == null`).
 *
 * **Privacy (ADR-0009):** [Request.params], [Response.result], [StreamData.event] and [Push.payload]
 * are opaque [JsonElement]s the transport never interprets — and **never renders in [toString]**
 * (redacted), so the privacy-critical [[Transcript]] frames can't leak through a stray log. Whether a
 * given payload *is* sensitive is unknowable to the transport, so it treats **all** payloads as such.
 */
@Serializable
sealed interface SidecarFrame {

    /**
     * Client→server handshake: presents the in-band auth [token] (the client half of peer-auth,
     * ADR-0024) and the [protocolVersion] the client speaks. The kernel uid check (`getpeereid`/
     * `SO_PEERCRED`) is the Helper's job; the client additionally verifies socket-path ownership +
     * perms out-of-band before sending this (see `PeerTrust`).
     */
    @Serializable
    @SerialName("hello")
    data class Hello(val token: String, val protocolVersion: Int) : SidecarFrame {
        override fun toString(): String = "Hello(token=<redacted>, protocolVersion=$protocolVersion)"
    }

    /**
     * Server→client handshake ack: the [protocolVersion] the Helper speaks and the [capabilities] it
     * advertises (D4) — the client surfaces these so consumers degrade gracefully against a Helper that
     * lacks a capability rather than calling a method that will only ever fail.
     */
    @Serializable
    @SerialName("welcome")
    data class Welcome(
        val protocolVersion: Int,
        val capabilities: Set<String> = emptySet(),
    ) : SidecarFrame

    /** Client→server: invoke [method] with opaque [params]. Opens a unary call or a server stream. */
    @Serializable
    @SerialName("request")
    data class Request(val id: Long, val method: String, val params: JsonElement? = null) : SidecarFrame {
        override fun toString(): String = "Request(id=$id, method=$method, params=${redact(params)})"
    }

    /** Server→client: the unary success [result] correlated to [Request.id]. */
    @Serializable
    @SerialName("response")
    data class Response(val id: Long, val result: JsonElement? = null) : SidecarFrame {
        override fun toString(): String = "Response(id=$id, result=${redact(result)})"
    }

    /** Server→client: one item of the server stream opened by [Request.id]. */
    @Serializable
    @SerialName("stream_data")
    data class StreamData(val id: Long, val event: JsonElement) : SidecarFrame {
        override fun toString(): String = "StreamData(id=$id, event=<redacted>)"
    }

    /** Server→client: the server stream opened by [Request.id] completed normally. */
    @Serializable
    @SerialName("stream_end")
    data class StreamEnd(val id: Long) : SidecarFrame

    /** Client→server: stop the server stream opened by [Request.id] (the consumer cancelled). */
    @Serializable
    @SerialName("cancel")
    data class Cancel(val id: Long) : SidecarFrame

    /** Server→client: an unsolicited event on [topic] (e.g. a permission change) with no correlation id. */
    @Serializable
    @SerialName("push")
    data class Push(val topic: String, val payload: JsonElement) : SidecarFrame {
        override fun toString(): String = "Push(topic=$topic, payload=<redacted>)"
    }

    /**
     * Server→client error. [id] correlates it to a [Request] (its response/stream failed); a null [id]
     * is a **connection-level** failure (e.g. a rejected [Hello]).
     */
    @Serializable
    @SerialName("failure")
    data class Failure(val id: Long? = null, val error: SidecarError) : SidecarFrame
}
